core/common/src/main/java/alluxio/resource/LockResource.java

```diff
@@ -70,6 +72,41 @@ public LockResource(Lock lock, boolean acquireLock, boolean useTryLock) {
     }
   }
 
+  /**
+   * Creates a new instance of {@link LockResource} using the given lock.
+   *
+   * This method will use the {@link Lock#tryLock(long, TimeUnit)} internally in a loop
+   * for trying to grab the lock without causing total blockage of other waiters of the lock.
+   *
+   * @param lock  the lock to acquire
+   * @param tryMs the duration to attempt acquiring the lock
+   * @param sleepMs the wait duration after failed attempt to acquire the lock
+   * @param timeoutMs the total duration to try acquiring the lock in a loop
+   * @throws InterruptedException if interrupted while acquiring the lock
+   * @throws TimeoutException if the lock was not acquired after given timeout
+   */
+  public LockResource(Lock lock, long tryMs, long sleepMs, long timeoutMs)
+      throws InterruptedException, TimeoutException {
+    mLock = lock;
+    long deadlineMs = System.currentTimeMillis() + timeoutMs;
+    boolean lockAcquired = false;
+    while (System.currentTimeMillis() < deadlineMs) {
+      if (mLock.tryLock(tryMs, TimeUnit.MILLISECONDS)) {
+        lockAcquired = true;
+        break;
+      } else {
+        long remainingWaitMs = deadlineMs = System.currentTimeMillis();
```

Comment:
is this supposed to be `deadlineMs - System.currentTimeMillis()`?

Reply:
yeah, it's the remaining wait till the timeout.

Reply:
nice catch! otherwise, it would always wait at the end and we'd never know.